src/pybind/mgr/k8sevents/module.py

```diff
@@ -0,0 +1,1029 @@
+                    event.count,
+                    event.message,
+                    event_name
+            )
+
+        return 0, "", s
+
+    def show_ceph_events(self):
+        if len(self.event_processor.events.keys()) > 0:
+            s = "\n".join(self.event_processor.events.keys())
+        else:
+            s = "No events\n"
+        return 0, "", s
+
+    def handle_command(self, inbuf, cmd):
+        # TODO Should we implement dynamic options for the monitoring?
+        if cmd["prefix"] == "k8sevents namespace":
+            return self.show_all_events()
+        elif cmd["prefix"] == "k8sevents ceph":
+            return self.show_ceph_events()
+        else:
+            raise NotImplementedError(cmd["prefix"])
+
+    @staticmethod
+    def can_run():
+        """Determine whether the pre-reqs for the module are in place"""
+
+        if not kubernetes_imported:
+            return False, "kubernetes python client is unavailable"
+        return True, ""
+
+    def log_callback(self, arg, line, channel, name, who, stamp_sec, stamp_nsec, seq, level, msg):
+        """Receive the ceph log entry and add it to the global queue"""
+
+        if sys.version_info[0] >= 3:
+            channel = channel.decode('utf-8')
+
+        # self.log.error("qsize={},line={},channel={},name={},who={},"
+        #                "stamp_sec={},seq={},level={},msg={}".format(event_queue.qsize(),line,channel,name,who,stamp_sec,
+        #                seq,level,msg))
+        
+        event_queue.put(
+            LogEntry(
+                source='log',
+                msg_type=channel,
+                msg=line.decode('utf-8'),
+                level=level[1:-1].decode('utf-8'),
+                tstamp=stamp_sec
+            )
+        )
+
+    def watch_ceph_log(self):
+        """Initiate the log_monitor2 call to mimic ceph -w behavior"""
+
+        level = self.get_localized_module_option(
+            'log_level', DEFAULT_LOG_LEVEL)
+
+        # create a daemon thread for monitor_log2
+        run_in_thread(self.rados.monitor_log2, level, self.log_callback, 0)
+
+    def serve(self):
+
+        self.event_processor = EventProcessor(logger=self.log)
+        self.config_watcher = CephConfigWatcher(self)
+        self.ns_watcher = NamespaceWatcher(logger=self.log)
+
+        if self.event_processor.ok:
+            self.log.warning("[INF] Ceph Log processor thread starting")
+            self.event_processor.start()        # start log consumer thread
+            self.log.warning("[INF] Ceph log watcher thread starting")
+            self.watch_ceph_log()               # start the log producer thread
+            self.log.warning("[INF] Ceph config watcher thread starting")
+            self.config_watcher.start()
+            self.log.warning("[INF] Rook-ceph namespace events watcher starting")
+            self.ns_watcher.start()
+
+            while True:
+                # stay alive
+                time.sleep(0.5)
```

Comment:
Have you tried not using `serve()` but the module's `__init__` to start those watchers? Because we're not really using this thread at all here.

Reply:
I think there is some code needed to check thread health. For example if the threads generating the events stop for some reason, I was going to send a error level message to the log. What do you think ?

Reply:
threads are now checked and reported in as failed when detected. The status command shows overall thread health for each of he trackers.